src/serialize.rs

```diff
@@ -0,0 +1,391 @@
+                        let expected = self.map_type_to_expected()?;
+                        return Err(SerializationError::new(mk_typck_err::<PyList>(
+                            typ,
+                            BuiltinTypeCheckErrorKind::MismatchedType { expected },
+                        )));
+                    }
+                };
+                dict.serialize(typ, cell_writer)
+            }
+            _ => unimplemented!("other types not supported yet"),
+        }
+    }
+}
+
+impl<'py> SerializeValue for PyAnyWrapper<'py> {
+    fn serialize<'a>(
+        &self,
+        typ: &ColumnType,
+        cell_writer: CellWriter<'a>,
+    ) -> Result<WrittenCellProof<'a>, SerializationError> {
+        if self.is_none() {
+            return Ok(cell_writer.set_null());
+        }
+
+        self.serialize_types(typ, cell_writer)
+    }
+}
+
+pub(crate) struct PyAnyWrapperRow(pub(crate) Py<PyAny>);
+
+impl Deref for PyAnyWrapperRow {
+    type Target = Py<PyAny>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl SerializeRow for PyAnyWrapperRow {
+    fn serialize(
+        &self,
+        ctx: &RowSerializationContext<'_>,
+        row_writer: &mut RowWriter,
+    ) -> Result<(), SerializationError> {
+        Python::with_gil(|py| {
+            let val = self.bind(py);
+
+            if let Ok(py_tuple) = val.downcast::<PyTuple>() {
+                if py_tuple.len() != ctx.columns().len() {
+                    return Err(SerializationError::new(mk_typck_err_row::<PyTuple>(
+                        scylla::serialize::row::BuiltinTypeCheckErrorKind::WrongColumnCount {
+                            rust_cols: py_tuple.len(),
+                            cql_cols: ctx.columns().len(),
+                        },
+                    )));
+                }
+
+                for (col, val) in ctx.columns().iter().zip(py_tuple.iter()) {
+                    let wrapper = PyAnyWrapper(val);
+                    let sub_writer = row_writer.make_cell_writer();
+                    SerializeValue::serialize(&wrapper, col.typ(), sub_writer)?;
+                }
+                return Ok(());
+            }
+
+            if let Ok(py_list) = val.downcast::<PyList>() {
+                if py_list.len() != ctx.columns().len() {
+                    return Err(SerializationError::new(mk_typck_err_row::<PyList>(
+                        scylla::serialize::row::BuiltinTypeCheckErrorKind::WrongColumnCount {
+                            rust_cols: py_list.len(),
+                            cql_cols: ctx.columns().len(),
+                        },
+                    )));
+                }
+
+                for (col, val) in ctx.columns().iter().zip(py_list.iter()) {
+                    let wrapper = PyAnyWrapper(val);
+                    let sub_writer = row_writer.make_cell_writer();
+                    SerializeValue::serialize(&wrapper, col.typ(), sub_writer)?;
+                }
+                return Ok(());
```

Comment:
There is quite a lot of duplication between the two cases. Is there no way to avoid that?